lib/extensions/attachable/active_record/base.rb

```diff
@@ -86,55 +145,128 @@ def files=(files)
+        id = parse_attachment_reference_uuid_from_url(image['src']) if image['src']
+        valid_id = get_valid_attachment_reference(id) if id
+
+        next unless valid_id
+        image['src'] = "#{ATTACHMENT_URL_PREFIX}#{valid_id}" unless valid_id == id
+        ids << valid_id
+      end
+
+      ids
+    end
+
+    # Parse all attachment_reference uuids in the content.
+    #
+    # @param [String] content The content which associated with the attachments.
+    # @return [Array<Integer>] the ids of the attachment references in the content.
+    def parse_attachment_reference_uuids_from_content(content)
+      ids = []
+      doc = Nokogiri::HTML(content)
+      doc.css('img').each do |image|
+        id = parse_attachment_reference_uuid_from_url(image['src'])
         ids << id if id
       end
 
       ids
     end
 
-    # Regex for filtering UUIDs.
+    # Regex for filtering Attachment IDs from URLs.
     ATTACHMENT_ID_REGEX = /\/attachments\/([0-9a-f]{8}-([0-9a-f]{4}-){3}[0-9a-f]{12})$/
-    # Parse attachment_reference from the given url.
+
+    # Parse attachment_reference uuid from the given url.
     #
-    # @param [String] url The url.
-    # @return [Integer|nil] the id of the attachment references in the url, nil will be returned
-    #   if the url is not a valid attachment url.
-    def parse_attachment_reference_id_from_url(url)
-      # TODO: Attachments from a third party domain with the same path should not be returned.
+    # @param [String] uuid The uuid.
+    # @return [String|nil] the uuid of the attachment_references, or nil if invalid.
+    def parse_attachment_reference_uuid_from_url(url)
       result = url.match(ATTACHMENT_ID_REGEX)
       result ? result[1] : nil
     end
+
+    # Gets the id of the attachment_reference associated with +self+. If an
+    # attachment_reference referencing a different attachable is provided, this method
+    # creates a new attachment_reference with the given attachment and name. If an
+    # incorrect UUID was provided, +nil+ is returned.
+    #
+    # This handles two cases:
+    #   i) Malformed attachment_reference id: returns +nil+.
+    #   ii) During duplication, or Copy/Paste: new attachment_references are
+    #       automatically created
+    #
+    # @param [String] id The given UUID of the attachment_reference
+    # @return [String|nil] nil if provided ID is not found, otherwise the UUID of the validated
+    #                      attachment_reference
+    def get_valid_attachment_reference(id)
+      object = AttachmentReference.find_by(id: id)
+      return nil unless object
+      return id if object.attachable == self || object.attachable.nil?
+
+      AttachmentReference.create(attachment: object.attachment, name: object.name).id
+    end
+
+    # Given the rich-text content, transform the src of image nodes to direct URLs.
+    #
+    # If S3 is used as a storage with signed URLs, this would return a URL that has an
```

Comment:
For future reference, this can be configured in the carrier_wave initializer.

https://github.com/carrierwaveuploader/carrierwave/blob/master/lib/carrierwave/storage/fog.rb#L47